rewrite-java/src/main/java/org/openrewrite/java/ChangePackage.java

```diff
@@ -113,14 +116,48 @@ public TreeVisitor<?, ExecutionContext> getVisitor() {
                     }
                     stopAfterPreVisit();
                 }
+                if (tree instanceof SourceFileWithTypeReferences) {
+                    SourceFileWithTypeReferences cu = (SourceFileWithTypeReferences) tree;
+                    boolean recursive = Boolean.TRUE.equals(ChangePackage.this.recursive);
+                    String recursivePackageNamePrefix = oldPackageName + ".";
+                    for (TypeReference ref : cu.getTypeReferences().getTypeReferences()) {
+                        if (ref.getName().equals(oldPackageName) || recursive && ref.getName().startsWith(recursivePackageNamePrefix)) {
```

Comment:
Ideally I think we should only be using the `matches()` and `renameTo()` methods instead of the `getName()` method, but coming up with a good API for those two methods feels pretty difficult, so maybe this will have to wait. I was thinking something along the lines of Java's `Matcher` for regex:

<img width="669" alt="image" src="https://github.com/user-attachments/assets/cb05b54c-8b19-442b-a7fd-c245c39363ce">

Reply:
One example I keep thinking about here is when we have `META-INF/services` files as we have them for the `TypeReference.Provider` interface. In this case the name of a nested type looks like `foo.Bar$Baz`. Now `ChangeType` can't possibly know that it needs to call `renameTo("foo.Bar$Baz")` and not `renameTo("foo.Bar.Baz")`. The latter is probably correct for some other config files. Currently, the `ChangeType` visitor would have to make sure to always pass in names like `"foo.Bar$Baz"` and then the `TypeReference` implementation would have to know if it should apply a heuristic (although `Bar$Baz` could even be the name of the top-level class, but that would be highly unlikely). The question is just what should the API look like to cater for this?